src/main/java/org/jenkinsci/plugins/workflow/graph/StandardGraphLookupView.java

```diff
@@ -24,50 +25,62 @@
 
     static final String INCOMPLETE = "";
 
-    /** Map the blockStartNode to its endNode, to accellerate a range of operations */
+    /** Map the blockStartNode to its endNode, to accelerate a range of operations */
     HashMap<String, String> blockStartToEnd = new HashMap<>();
 
     /** Map a node to its nearest enclosing block */
     HashMap<String, String> nearestEnclosingBlock = new HashMap<>();
 
+    final ReentrantLock lock = new ReentrantLock();
+
     public void clearCache() {
-        blockStartToEnd.clear();
-        nearestEnclosingBlock.clear();
+        lock.lock();
+        try {
+            blockStartToEnd.clear();
+            nearestEnclosingBlock.clear();
+        } finally {
+            lock.unlock();
+        }
     }
 
     /** Update with a new node added to the flowgraph */
     @Override
     public void onNewHead(@Nonnull FlowNode newHead) {
-        if (newHead instanceof BlockEndNode) {
-            blockStartToEnd.put(((BlockEndNode)newHead).getStartNode().getId(), newHead.getId());
-            String overallEnclosing = nearestEnclosingBlock.get(((BlockEndNode) newHead).getStartNode().getId());
-            if (overallEnclosing != null) {
-                nearestEnclosingBlock.put(newHead.getId(), overallEnclosing);
-            }
-        } else {
-            if (newHead instanceof BlockStartNode) {
-                blockStartToEnd.put(newHead.getId(), INCOMPLETE);
-            }
-
-            // Now we try to generate enclosing block info for caching, by looking at parents
-            // But we don't try to hard -- usually the cache is populated and we defer recursive walks of the graph
-            List<FlowNode> parents = newHead.getParents();
-            if (parents.size() > 0) {
-                FlowNode parent = parents.get(0);  // Multiple parents only for end of a parallel, and then both are BlockEndNodes
+        lock.lock();
+        try {
+            if (newHead instanceof BlockEndNode) {
+                blockStartToEnd.put(((BlockEndNode) newHead).getStartNode().getId(), newHead.getId());
+                String overallEnclosing = nearestEnclosingBlock.get(((BlockEndNode) newHead).getStartNode().getId());
```

Comment:
The sort of thing that makes me dubious you could a lockless data structure for this purpose.

Reply:
@dwnusbaum suggests that you probably could, if you wanted to review the logic carefully.